### PR TITLE
#4521 Replace native errors with @fastify/errors

### PR DESCRIPTION
**Pull request by alainrk**
This PR addresses https://github.com/fastify/fastify/issues/4521 by replacing native errors (and where needed generalizing them) with `@fastify/errors`.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Reference/Errors.md [hunk 1]
```diff
@@ -133,6 +133,41 @@ fastify.listen({ port: 3000 }, function (err, address) {
...
+
+#### FST_ERR_AJV_CUSTOM_OPTIONS_OPT_NOT_OBJ
+<a id="FST_ERR_AJV_CUSTOM_OPTIONS_OPT_NOT_OBJ"></a>
+
+Sajv.customOptions option should be an object.
+
+#### FST_ERR_AJV_CUSTOM_OPTIONS_OPT_NOT_ARR
+<a id="FST_ERR_AJV_CUSTOM_OPTIONS_OPT_NOT_ARR"></a>
+
+Sajv.plugins option should be an array.
```
Comment (Uzlopak) [inline docs/Reference/Errors.md:164]:
```suggestion
Sajv.plugins option should be an array.
```

File: test/internals/reply.test.js [hunk 2]
```diff
@@ -1740,7 +1740,7 @@ test('cannot set the replySerializer when the server is running', t => {
       fastify.setReplySerializer(() => {})
       t.fail('this serializer should not be setup')
     } catch (e) {
-      t.equal(e.message, 'Cannot call "setReplySerializer" when fastify instance is already started!')
+      t.equal(e.message, 'Fastify is already listening. Cannot call "setReplySerializer"!')
```
Comment (mcollina) [inline test/internals/reply.test.js:1743]: Please replate this assertion with one that verifies the error code.

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work!

File: test/validation-error-handling.test.js [hunk 3]
```diff
@@ -568,22 +568,22 @@ test('cannot create a fastify instance with wrong type of errorFormatter', t =>
...
-    t.equal(err.message, 'schemaErrorFormatter option should be a function, instead got number')
+    t.equal(err.message, "schemaErrorFormatter option should be a non async function. Instead got 'number'.")
   }
 
   try {
     const fastify = Fastify()
     fastify.setSchemaErrorFormatter(500)
   } catch (err) {
-    t.equal(err.message, 'schemaErrorFormatter option should be a function, instead got number')
+    t.equal(err.message, "schemaErrorFormatter option should be a non async function. Instead got 'number'.")
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:586]:
```suggestion
    t.equal(err.code, 'FST_ERR_SCHEMA_ERROR_FORMATTER_NOT_FN')
```

File: test/validation-error-handling.test.js [hunk 4]
```diff
@@ -568,22 +568,22 @@ test('cannot create a fastify instance with wrong type of errorFormatter', t =>
...
+    t.equal(err.message, "schemaErrorFormatter option should be a non async function. Instead got 'AsyncFunction'.")
   }
 
   try {
     Fastify({
       schemaErrorFormatter: 500
     })
   } catch (err) {
-    t.equal(err.message, 'schemaErrorFormatter option should be a function, instead got number')
+    t.equal(err.message, "schemaErrorFormatter option should be a non async function. Instead got 'number'.")
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:579]:
```suggestion
    t.equal(err.code, 'FST_ERR_SCHEMA_ERROR_FORMATTER_NOT_FN')
```

File: test/validation-error-handling.test.js [hunk 5]
```diff
@@ -568,22 +568,22 @@ test('cannot create a fastify instance with wrong type of errorFormatter', t =>
       }
     })
   } catch (err) {
-    t.equal(err.message, 'schemaErrorFormatter option should not be an async function')
+    t.equal(err.message, "schemaErrorFormatter option should be a non async function. Instead got 'AsyncFunction'.")
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:571]:
```suggestion
    t.equal(err.code, 'FST_ERR_SCHEMA_ERROR_FORMATTER_NOT_FN')
```

File: test/internals/reply.test.js (hunk 2 above)
Comment (Uzlopak) [reply test/internals/reply.test.js:1743]: t.equal(e.code, 'FST_ERR_REOPENED_SERVER')

File: test/validation-error-handling.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:571]: @mcollina replace or add?

File: test/validation-error-handling.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:586]: @mcollina replace or add?

File: test/validation-error-handling.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Uzlopak) [reply test/validation-error-handling.test.js:579]: @mcollina replace or add?

File: fastify.js [hunk 6]
```diff
@@ -463,7 +473,7 @@ function fastify (options) {
   return fastify
 
   function throwIfAlreadyStarted (msg) {
-    if (fastify[kState].started) throw new Error(msg)
+    if (fastify[kState].started) throw new FST_ERR_REOPENED_SERVER(msg)
```
Comment (Uzlopak) [inline fastify.js:476]: I am not happy with this. It should be something like `FST_ERR_INSTANCE_ALREADY_LISTENING` or so.

Comment (alainrk) [reply fastify.js:476]: I agree, I think I've reused the existing one with the same semantic, are you ok with renaming it across the board? (mentioned in this comment: https://github.com/fastify/fastify/issues/4521#issuecomment-1406252081)

File: lib/errors.js [hunk 7]
```diff
@@ -266,7 +358,7 @@ const codes = {
   ),
   FST_ERR_REOPENED_SERVER: createError(
     'FST_ERR_REOPENED_SERVER',
-    'Fastify is already listening'
+    'Fastify is already listening. %s'
```
Comment (Uzlopak) [inline lib/errors.js:361]: Revert this

File: fastify.js (hunk 6 above)
Comment (Uzlopak) [reply fastify.js:476]: Renaming it means that it is a breaking change. Create a new error and use it where we used native errors.
See also below comment

Comment (alainrk) [reply fastify.js]: Sure no problem, pushed now 👍

File: test/validation-error-handling.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (alainrk) [reply test/validation-error-handling.test.js]: I don't like much the check on the message if the code is available, but I can put it back if needed @Uzlopak @mcollina

File: test/validation-error-handling.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply test/validation-error-handling.test.js]: add would be better, thx!

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**